storage/innobase/srv/srv0start.cc

```diff
@@ -1994,6 +1994,11 @@ void innodb_preshutdown()
   srv_shutdown_bg_undo_sources();
   srv_purge_shutdown();
 
+  /* Do system tablespace truncation during slow shutdown */
+  if (!srv_fast_shutdown && !high_level_read_only && srv_was_started
+      && srv_operation == SRV_OPERATION_NORMAL)
+    fsp_system_tablespace_truncate();
```

Comment:
It is unclear to me how things are improved when this piece of code is executed _earlier_. The pull request mentions that the test `innodb.sys_truncate_shutdown` covers this code, but the test was not extended to demonstrate the impact of this. Is there any (even nondeterministic) way of demonstrating the impact?

Lines should customarily be split _after_, not _before_ binary operators, such as `&&`. Only the archaic InnoDB formatting style followed that convention.

Reply:
I now see that the original code was misplaced, invoked before `innodb_shutdown()`, which would invoke `logs_empty_and_mark_files_at_shutdown()`, which would seem to be a safer place to invoke the code.

An even more natural place for this code might be in `srv_purge_shutdown()`, which is handling `!srv_fast_shutdown` as a special case due to the logic in `srv_purge_should_exit()`.